More wide character fixes and cleanup.

mahge opened the pull request:
- Implement wide character handling for `SystemImpl__directoryExists`
    and `System_subDirectories`.

  - Remove the macros for conversion to/from wide character string, e.g,
    `MULTIBYTE_TO_WIDECHAR_LENGTH`, `MULTIBYTE_TO_WIDECHAR_VAR` ...
    and replace them with corresponding functions.

  - Added functions `omc_multibyte_to_wchar_str` and `omc_wchar_to_multibyte_str`
    to handle the conversions.
    Note: the caller has to deallocated the memory for the strings returned
    from these functions.

  - define `omc_stat_t` to be `stat` or `_stat` depending on platform and use
    it instead. We might need to change this to wstat on Windows later.

  - A few more minor cleanups and restructures.

  - Fixes #8777.

AnHeuermann reviewed (commented):
Looking good.

There are still a lot of places using `struct stat` instead of the new `omc_stat_t` in systemimpl.c. Should they be replaced in this PR?

mahge commented:
> There are still a lot of places using struct stat instead of the new omc_stat_t in systemimpl.c. Should they be replaced in this PR?

Good point. I just replaced the ones in the places I modified. I will replace them all.
